src/sipmpdf/functions.py

```diff
@@ -216,47 +245,384 @@ def darkcurrent_response(x: np.float64,
-  poisson_mean: np.float64,
-  poisson_borel: np.float64,
-  ap_prob: np.float64,
-  ap_beta: np.float64,
-) -> np.float64:
+  """
+  Dark current response with random noise, given a detector resolution factor
+  (calculated relative to the gain of the system)
+
+  Parameters
+  ----------
+  x : np.float64
+      Observation
+  smear : np.float64
+      The value by which the dark current response is smeared
+  gain : np.float64
+      The gain of the SiPM
+  resolution : np.float64, optional
+      resolution factor, relative to the SiPM gain factor, by default 1e-4
+
+  Returns
+  -------
+  np.float64
+      Probability density of observing a dark current response
+  """
+
+  kern = kernel_switch(x, smear, gain, resolution)
+
+  n_samples = 1024
+
+  #extend arrays
+  def extend_arr(x):
+    return kern.repeat(x[kern.newaxis, ...], n_samples, axis=0)
+
+  x = extend_arr(x)
+  smear = extend_arr(smear)
+  gain = extend_arr(gain)
+  resolution = extend_arr(resolution)
+
+  delta = 8 * smear / n_samples  # Getting the integration window
+  idx = kern.local_index(x, axis=0)
+  x_pr = -4 * smear + delta * (idx + 0.5)
+
+  # Given that the computed result will exhibit periodicity within the
+  # integration window, here we compute a simple rolling means with 5 parts
+  return kern.sum(darkcurrent_response(x - x_pr, gain, resolution) *
+                  normal(x_pr, 0, smear) * delta,
+                  axis=0)
+
+
+@expand_shape
+def afterpulsing_summation(x: np.float64, total: np.int64, ap_smear: np.float64,
+                           ap_beta: np.float64,
+                           ap_prob: np.float64) -> np.float64:
+  """
+  Summation of the afterpulse response weighted by the binomial distribution for n_ap from 1 to 10
+  
+  Parameters
+  ----------
+  x : np.float64
+      Observation
+  ap_smear: np.float64
+      Value by which the afterpulse response is smeared by
+  ap_beta : np.float64
+      Afterpulse timescale factor.
+  Returns
+  -------
+  np.float64
+      Afterpulse response weighted by binomial disribution
+  """
+
+  kern = kernel_switch(x, ap_smear, ap_prob, ap_beta, total)
+  k = 10
+
+  #extend arrays
+  def extend_arr(x):
+    return kern.repeat(x[kern.newaxis, ...], k, axis=0)
+
+  x = extend_arr(x)
+  ap_smear = extend_arr(ap_smear)
+  ap_beta = extend_arr(ap_beta)
+  ap_prob = extend_arr(ap_prob)
+  total = extend_arr(total)
+
+  idx = kern.local_index(x, axis=0)
+  idx += 1
+
+  return kern.sum(
+    binomial_prob(x=idx, total=total, prob=ap_prob) *
+    ap_response_smeared(x=x, smear=ap_smear, n_ap=idx, beta=ap_beta),
+    axis=0)
+
+
+@expand_shape
+def k_summation(x: np.float64, common_noise: np.float64, pixel_noise: np.float64,
+                ap_beta: np.float64, ap_prob: np.float64, pedestal: np.float64,
+                gain: np.float64, poisson_mean: np.float64,
+                poisson_borel: np.float64) -> np.float64:
+  """ 
+  Summation in the final SiPM response for k>=1 not accounting for dark current
```

Comment:
"..for number of primary P.E.s k>=1, dark current effect at k==0 is handled in a separate function"